core/src/smartcontracts/wasm.rs

```diff
@@ -622,6 +701,27 @@ trait ExecuteOperations<'wrld, S: state::GetCommon<'wrld>> {
     }
 }
 
+trait QueryAuthority<'wrld, S> {
+    #[codec::wrap_trait_fn]
+    fn query_authority(state: &S) -> AccountId;
+}
+
+/// Marker trait to have [`QueryAuthority`] default-implemented for concrete [`Runtime`]
+trait DefaultQueryAuthority {}
+
+impl<'wrld, S: state::GetCommon<'wrld>, R: DefaultQueryAuthority> QueryAuthority<'wrld, S> for R {
```

Comment:
```suggestion
impl<'wrld, S: state::GetCommon<'wrld> + HasAuthority> QueryAuthority<'wrld, S> for Runtime<S> {
```

can it be like this?

Reply:
it's not public api so I won't insist on this

Reply:
> Why do we need `QueryAuthority` trait at all?

Because I can't implement it of `Runtime<S>` as `Runtime<state::ValidatorPermisisonTokens>` have another implementation for it and that would be a mulitple definition of the same function.

And because of that I can't do what you propose here, but I can implement the idea with `HasAuthority`

Reply:
I wasn't right, if we add constraints to `S` when we can implement it on `Runtime<S>`. Check out new version